Backport py3.11 asyncio's taskgroups. (#8791)
micropython/micropython: docs/library/asyncio.rst

    ignore this exception.  Cleanup code may be run by trapping it, or via
    ``try ... finally``.

> class Taskgroup

Author: Right. Fixing.

---------------

> 

Reviewer: At the moment the link at the top of this page is to CPython 3.8's asyncio reference, which doesn't have TaskGroups. Suggest starting this section with a short explanation of this being a port of the CPython 3.11 feature, and a link to https://docs.python.org/3.11/library/asyncio-task.html#task-groups

See Nathaniel J. Smith's `essay on Structured Concurrency
<https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/>`_
for an introduction why you should use taskgroups instead of starting
"naked" tasks.

.. note::
    His "nursery" objects are called "taskgroup" in asyncio; the

>     equivalent of a "go statement" is `Loop.create_task`.

Reviewer: This is interesting context and motivation, but I'm a bit torn on whether this is suitable for the MicroPython asyncio reference or not. (i.e. the equivalent CPython docs provide a single sentence summary of what TaskGroups do and that feels more appropriate for reference documentation.)

.. class:: TaskGroup()

>     This object is an async context managed holding a group of tasks.
>     Tasks can be added to the group using `TaskGroup.create_task`.

Reviewer: ```suggestion
    An asynchronous context manager holding a group of tasks. Tasks can be
    added to the group using `TaskGroup.create_task`. All tasks are awaited when
    the context manager exits.
```

(Originally this was just a suggestion to fix the typo, but changed it to copying the CPython doc summary exactly!)

    If a task belonging to the group fails, the remaining tasks in the
    group are cancelled with an :exc:`asyncio.CancelledError` exception.
    (This also holds for the code within the context manager's block.)
    No further tasks can then be added to the group.

    When there is no exception, leaving the context manager waits for
    the taskgroup's member tasks to end before proceeding. It does not
    cancel these tasks and does not prevent the creation of new tasks.

.. method:: TaskGroup.create_task(coroutine)

    Create a subtask that executes *coroutine* as part of this taskgroup.

    Returns the new task.

.. method:: TaskGroup.cancel()

    Stop the taskgroup, i.e. cancel all its tasks.

    This method is equivalent to cancelling the task responsible for the
    body of the taskgroup, *if* that is what the task is currently doing.

> 

Reviewer: ```suggestion
    .. note:: This is a MicroPython-only method.
```

(Similarly, if there's anywhere else here that a function or behaviour is different to CPython then please add an explanatory note.)

~~I noticed that CPython's solution to this has been added to the docs in recent versions:
https://docs.python.org/3.14/library/asyncio-task.html#terminating-a-task-group~~

~~Curious if you think that's a viable alternative to having a `TaskGroup.cancel()` method?~~ EDIT: nevermind!

Reviewer: > Curious if you think that's a viable alternative to having a `TaskGroup.cancel()` method?

Oh, I missed your comment that CPython is getting this function! In which case maybe can say that here instead (that it's a CPython 3.15...? feature?)

> .. exception:: Cancelled

Reviewer: This exception doesn't seem to exist in CPython, it shares `asyncio.CancelledError` for this case - doesn't it?

> .. exception:: Cancelled
> 
>     This exception is raised in a task whose taskgroup is being cancelled.
> 
>     This is a subclass of ``BaseException``; it should never be caught.

Reviewer: ```suggestion
.. exception:: CancelledError

   The operation has been cancelled.

   This exception can be caught to perform custom operations when asyncio Tasks are cancelled. In almost all situations the exception must be re-raised.

    This is a subclass of ``BaseException``.
```

Somehow `CancelledError` wasn't documented in MicroPython until now! The above suggestion is cribbed from the CPython docs, but I think their advice to catch and re-raise is more useful than blanket "should never be caught".

(There's a follow-up task for someone to change all the occurences of `asyncio.CancelledError` in this doc to be reference links, but you don't have to do that chore in this PR!)

.. exception:: ExceptionGroup

    If multiple subtasks raise exceptions in parallel, it's unclear which
    of them should be propagated. Thus an `ExceptionGroup` exception
    collects them and is raised instead.

> 

Reviewer: ```suggestion
    .. note:: In CPython ExceptionGroup is a built-in exception, but in MicroPython it is
       part of the asyncio module.
        
       MicroPython does not support CPython 3.11's ``except*`` syntax for handling
       exception groups automatically.
```

.. method:: ExceptionGroup.split(typ)

    This method can be used to filter the exceptions within an exception group.
    It returns two lists: the first contains those sub-exceptions which
    match *typ*, the second those which do not.

    *typ* can be an exception class, a list of exception classes, or a
    callable that returns ``True`` if the exception passed to it should be
    returned in the first list.

    MicroPython does not support CPython 3.11's syntax for filtering handling

>     exception groups.

Reviewer: ```suggestion
    MicroPython does not support CPython 3.11's ``except*`` syntax for handling
    exception groups without calling this function explicitly.
```

>     MicroPython does not support CPython 3.11's syntax for filtering handling
>     exception groups.
> 

Reviewer: ```suggestion
```

Suggest removing this from here and covering it in the class documentation above.

.. function:: start_server(callback, host, port, backlog=5, ssl=None)

    Start a TCP server on the given *host* and *port*.  For each incoming,
    accepted connection, *callback* will be called in a new task with
    2 arguments: reader and writer streams for the connection.

    This function does **not** co-operate well with taskgroups.
    If you use them, you should use a function like `run_server`
    (supplied in ``examples/run_server.py``) instead.

    If *ssl* is a `ssl.SSLContext` object, this context is used to create the transport.

    Returns a `Server` object.

    This is a coroutine.

> .. function:: run_server(callback, host, port, backlog=5, taskgroup=None)

Reviewer: I have two concerns here:

1. This function doesn't exist in CPython 3.11 asyncio, and we're trying not to add any more MicroPython-only names in CPython-compatible modules..
2. This function doesn't appear to exist in MicroPython asyncio module either, at least in this version of the PR it's in the `examples/run_server.py` file. Ergonomically this is a bit rough for someone who wants to use it.

Is the comment above about how `start_server` doesn't work well with TaskGroups exclusive to MicroPython? If so, is there any way that we can change MicroPython's implementation of `start_server` to be closer to CPython's in this regard?

(Otherwise, if `start_server` also doesn't work well with TaskGroups on CPython then a good path forward might be to add `run_server` as a micropython-lib library that can be separately `mip install`-ed, and mention it in the docs for `start_server` here.)

    Create a task from the given *coro* and return the new `Task` object.

>     You should not call this function when you're using taskgroups.

Reviewer: ```suggestion
    .. warning: Due to MicroPython internal limitations, this function can't be used at the same time as TaskGroups.
```
